apps/condo/domains/organization/utils/accessSchema.js

```diff
@@ -1,191 +1,297 @@
+ * can be passed as array of strings for multiple permissions or a single string for a single permission
+ * @returns {Promise<Array<string>>}
  */
-async function checkOrganizationsPermission (userId, organizationIds, permission) {
-    if (!userId || !isArray(organizationIds) || isEmpty(organizationIds)) return false
+async function getEmployedOrganizationsByPermissions (user, permissions) {
+    const userOrganizationsInfo = await _getUserOrganizations(user)
 
-    const uniqOrganizationIds = uniq(organizationIds)
+    const permissionsToCheck = Array.isArray(permissions) ? permissions : [permissions]
+    const organizationIds = []
 
-    if (!uniqOrganizationIds.every(isUUID)) return false
-
-    for (const id of organizationIds) {
-        const hasAccess = await checkOrganizationPermission(userId, id, permission)
-        if (!hasAccess) return false
+    for (const [id, info] of Object.entries(userOrganizationsInfo.organizations)) {
+        if (_checkPermissionsInOrganization(info.permissions, permissionsToCheck)) {
+            organizationIds.push(id)
+        }
     }
-
-    return true
+    return organizationIds
 }
 
 /**
- * Check that the user has access in each related organization
- *
- * @param userId {string} User.id field
- * @param organizationIds {Array<string>} array of objects related organizations
- * @param permission {string} OrganizationEmployeeRole permission key to check for
- * @return {Promise<boolean>}
+ * Gets the IDs of those child (related) organizations in whose parent organizations the user is employed and has all permissions from the list.
+ * Most likely you don't need to call this method directly, but use getEmployedOrganizationsByPermissions or getEmployedOrRelatedOrganizationsByPermissions
+ * @param {{ id: string }} user - user object
+ * @param {Array<string> | string} permissions - permissions to check,
+ * can be passed as array of strings for multiple permissions or a single string for a single permission
+ * @returns {Promise<Array<string>>}
  */
-async function checkRelatedOrganizationsPermission (userId, organizationIds, permission) {
-    if (!userId || !isArray(organizationIds) || isEmpty(organizationIds)) return false
-
-    const uniqOrganizationIds = uniq(organizationIds)
+async function getRelatedOrganizationsByPermissions (user, permissions) {
+    const userOrganizationsInfo = await _getUserOrganizations(user)
 
-    if (!uniqOrganizationIds.every(isUUID)) return false
+    const permissionsToCheck = Array.isArray(permissions) ? permissions : [permissions]
+    const organizationIds = []
 
-    for (const id of organizationIds) {
-        const hasAccess = await checkRelatedOrganizationPermission(userId, id, permission)
-        if (!hasAccess) return false
+    for (const info of Object.values(userOrganizationsInfo.organizations)) {
+        if (_checkPermissionsInOrganization(info.permissions, permissionsToCheck)) {
+            organizationIds.push(...info.childOrganizations)
```

Comment:
`Array.concat` will show better performance here

Reply:
R u sure? concat usually create new array of size arr1 + arr2, and copy the whole sequence...Why push is expanding current array

Reply:
https://dev.to/uilicious/javascript-array-push-is-945x-faster-than-array-concat-1oki

Reply:
Yeah, my bad. My test had an issue, so I get a wrong result